File: blocker.go
// blocker.go
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
	blue   = "\033[94m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Config struct {
	Blocked   []string `json:"blocked"`
	Active    bool     `json:"active"`
	TimerEnd  string   `json:"timer_end,omitempty"`
	Whitelist []string `json:"whitelist"`
}

func getConfigDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".blocker")
}

func getConfigFile() string {
	return filepath.Join(getConfigDir(), "config.json")
}

func getHostsFile() string {
	if runtime.GOOS == "windows" {
		return `C:\Windows\System32\drivers\etc\hosts`
	}
	return "/etc/hosts"
}

const blockIP = "127.0.0.1"

func ensureConfig() {
	os.MkdirAll(getConfigDir(), 0755)
	if _, err := os.Stat(getConfigFile()); os.IsNotExist(err) {
		cfg := Config{Blocked: []string{}, Active: false, Whitelist: []string{}}
		data, _ := json.MarshalIndent(cfg, "", "  ")
		ioutil.WriteFile(getConfigFile(), data, 0644)
	}
}

func loadConfig() Config {
	ensureConfig()
	data, err := ioutil.ReadFile(getConfigFile())
	if err != nil {
		return Config{}
	}
	var cfg Config
	json.Unmarshal(data, &cfg)
	return cfg
}

func saveConfig(cfg Config) {
	data, _ := json.MarshalIndent(cfg, "", "  ")
	ioutil.WriteFile(getConfigFile(), data, 0644)
}

func readHosts() ([]string, error) {
	file, err := os.Open(getHostsFile())
	if err != nil {
		return nil, err
	}
	defer file.Close()
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines, scanner.Err()
}

func writeHosts(lines []string) error {
	return ioutil.WriteFile(getHostsFile(), []byte(strings.Join(lines, "\n")), 0644)
}

func updateHosts(domains []string, action string) error {
	lines, err := readHosts()
	if err != nil {
		return err
	}
	var newLines []string
	for _, line := range lines {
		skip := false
		for _, domain := range domains {
			if strings.HasPrefix(strings.TrimSpace(line), blockIP) && strings.Contains(line, domain) {
				skip = true
				break
			}
		}
		if !skip {
			newLines = append(newLines, line)
		}
	}
	if action == "add" {
		for _, domain := range domains {
			newLines = append(newLines, fmt.Sprintf("%s %s", blockIP, domain))
		}
	}
	return writeHosts(newLines)
}

func startBlocking(domains []string, durationMinutes int) {
	cfg := loadConfig()
	cfg.Blocked = append(cfg.Blocked, domains...)
	if durationMinutes > 0 {
		cfg.TimerEnd = time.Now().Add(time.Duration(durationMinutes) * time.Minute).Format(time.RFC3339)
	}
	cfg.Active = true
	saveConfig(cfg)

	if err := updateHosts(domains, "add"); err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		os.Exit(1)
	}
	fmt.Println(colorize(fmt.Sprintf("Blocked %d domains.", len(domains)), green))
	if durationMinutes > 0 {
		fmt.Println(colorize(fmt.Sprintf("Blocking active for %d min.", durationMinutes), yellow))
		go timerThread(durationMinutes)
	}
}

func timerThread(minutes int) {
	time.Sleep(time.Duration(minutes) * time.Minute)
	fmt.Println(colorize("Time expired. Unblocking...", yellow))
	disableBlocking()
}

func disableBlocking() {
	cfg := loadConfig()
	blocked := cfg.Blocked
	if len(blocked) == 0 {
		fmt.Println(colorize("No active blocking.", yellow))
		return
	}
	if err := updateHosts(blocked, "remove"); err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		return
	}
	cfg.Blocked = []string{}
	cfg.Active = false
	cfg.TimerEnd = ""
	saveConfig(cfg)
	fmt.Println(colorize("Blocking disabled.", green))
}

func showStatus() {
	cfg := loadConfig()
	if cfg.Active && cfg.TimerEnd != "" {
		end, _ := time.Parse(time.RFC3339, cfg.TimerEnd)
		remaining := int(end.Sub(time.Now()).Seconds())
		if remaining > 0 {
			mins := remaining / 60
			secs := remaining % 60
			fmt.Println(colorize(fmt.Sprintf("Blocking active. Remaining: %d min %d sec", mins, secs), blue))
		} else {
			fmt.Println(colorize("Blocking active but timer expired.", yellow))
		}
	} else if cfg.Active {
		fmt.Println(colorize("Blocking active (no timer).", blue))
	} else {
		fmt.Println(colorize("Blocking inactive.", yellow))
	}
	if len(cfg.Blocked) > 0 {
		fmt.Println(colorize(fmt.Sprintf("Blocked domains (%d):", len(cfg.Blocked)), green))
		for _, d := range cfg.Blocked {
			fmt.Printf("  - %s\n", d)
		}
	} else {
		fmt.Println(colorize("No blocked domains.", yellow))
	}
}

func main() {
	var (
		addCmd      = flag.NewFlagSet("add", flag.ExitOnError)
		removeCmd   = flag.NewFlagSet("remove", flag.ExitOnError)
		listCmd     = flag.NewFlagSet("list", flag.ExitOnError)
		enableCmd   = flag.NewFlagSet("enable", flag.ExitOnError)
		disableCmd  = flag.NewFlagSet("disable", flag.ExitOnError)
		statusCmd   = flag.NewFlagSet("status", flag.ExitOnError)
		pomodoroCmd = flag.NewFlagSet("pomodoro", flag.ExitOnError)
		clearCmd    = flag.NewFlagSet("clear", flag.ExitOnError)
	)

	if len(os.Args) < 2 {
		fmt.Println(colorize("Usage: blocker <add|remove|list|enable|disable|status|pomodoro|clear> [domains...] [options]", yellow))
		os.Exit(1)
	}

	switch os.Args[1] {
	case "add":
		addCmd.Parse(os.Args[2:])
		domains := addCmd.Args()
		if len(domains) == 0 {
			fmt.Println(colorize("Specify domains to block.", yellow))
			return
		}
		startBlocking(domains, 0)

	case "remove":
		removeCmd.Parse(os.Args[2:])
		domains := removeCmd.Args()
		if len(domains) == 0 {
			fmt.Println(colorize("Specify domains to remove.", yellow))
			return
		}
		cfg := loadConfig()
		for _, d := range domains {
			for i, b := range cfg.Blocked {
				if b == d {
					cfg.Blocked = append(cfg.Blocked[:i], cfg.Blocked[i+1:]...)
					break
				}
			}
		}
		saveConfig(cfg)
		if err := updateHosts(domains, "remove"); err != nil {
			fmt.Println(colorize("Error: "+err.Error(), red))
		} else {
			fmt.Println(colorize(fmt.Sprintf("Removed: %s", strings.Join(domains, ", ")), green))
		}

	case "list":
		listCmd.Parse(os.Args[2:])
		cfg := loadConfig()
		blocked := cfg.Blocked
		if len(blocked) == 0 {
			fmt.Println(colorize("No blocked domains.", yellow))
		} else {
			fmt.Println(colorize("Blocked domains:", green))
			for _, d := range blocked {
				fmt.Printf("  - %s\n", d)
			}
		}

	case "enable":
		timeFlag := enableCmd.Int("t", 0, "Duration in minutes")
		enableCmd.Parse(os.Args[2:])
		cfg := loadConfig()
		if len(cfg.Blocked) == 0 {
			fmt.Println(colorize("No domains in blocklist. Add them with 'add'.", yellow))
			return
		}
		startBlocking(cfg.Blocked, *timeFlag)

	case "disable":
		disableCmd.Parse(os.Args[2:])
		disableBlocking()

	case "status":
		statusCmd.Parse(os.Args[2:])
		showStatus()

	case "pomodoro":
		pomodoroCmd.Parse(os.Args[2:])
		cfg := loadConfig()
		if len(cfg.Blocked) == 0 {
			fmt.Println(colorize("No domains in blocklist. Add them with 'add'.", yellow))
			return
		}
		fmt.Println(colorize("Starting Pomodoro: 25 minutes work", blue))
		startBlocking(cfg.Blocked, 25)
		time.Sleep(25 * time.Minute)
		fmt.Println(colorize("Pomodoro finished! 5 minutes break.", green))
		disableBlocking()
		time.Sleep(5 * time.Minute)
		fmt.Println(colorize("Break over. Start next Pomodoro.", blue))

	case "clear":
		force := clearCmd.Bool("f", false, "Force")
		clearCmd.Parse(os.Args[2:])
		if !*force {
			fmt.Print(colorize("Clear all blocklist? [y/N] ", yellow))
			var ans string
			fmt.Scanln(&ans)
			if ans != "y" && ans != "Y" {
				return
			}
		}
		cfg := loadConfig()
		if len(cfg.Blocked) == 0 {
			fmt.Println(colorize("Blocklist already empty.", yellow))
			return
		}
		if err := updateHosts(cfg.Blocked, "remove"); err != nil {
			fmt.Println(colorize("Error: "+err.Error(), red))
		} else {
			cfg.Blocked = []string{}
			cfg.Active = false
			saveConfig(cfg)
			fmt.Println(colorize("Blocklist cleared.", green))
		}

	default:
		fmt.Println(colorize("Unknown command: "+os.Args[1], red))
	}
}

File: blocker.cpp
// blocker.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <filesystem>
#include <thread>
#include <chrono>
#include <algorithm>
#include <json/json.h> // sudo apt-get install libjsoncpp-dev

using namespace std;
namespace fs = std::filesystem;

const string RESET = "\033[0m";
const string GREEN = "\033[92m";
const string RED = "\033[91m";
const string YELLOW = "\033[93m";
const string BLUE = "\033[94m";

string colorize(const string& text, const string& color) {
    return color + text + RESET;
}

string getConfigDir() {
    const char* home = getenv("HOME");
    if (!home) home = getenv("USERPROFILE");
    return string(home) + "/.blocker";
}

string getConfigFile() {
    return getConfigDir() + "/config.json";
}

string getHostsFile() {
#ifdef _WIN32
    return "C:\\Windows\\System32\\drivers\\etc\\hosts";
#else
    return "/etc/hosts";
#endif
}

const string BLOCK_IP = "127.0.0.1";

Json::Value loadConfig() {
    fs::create_directories(getConfigDir());
    ifstream f(getConfigFile());
    Json::Value root;
    if (!f) {
        root["blocked"] = Json::arrayValue;
        root["active"] = false;
        root["whitelist"] = Json::arrayValue;
        return root;
    }
    f >> root;
    return root;
}

void saveConfig(const Json::Value& root) {
    ofstream f(getConfigFile());
    f << root.toStyledString();
}

vector<string> readHosts() {
    ifstream f(getHostsFile());
    vector<string> lines;
    string line;
    while (getline(f, line)) lines.push_back(line);
    return lines;
}

void writeHosts(const vector<string>& lines) {
    ofstream f(getHostsFile());
    for (const auto& line : lines) f << line << endl;
}

void updateHosts(const vector<string>& domains, const string& action) {
    auto lines = readHosts();
    vector<string> newLines;
    for (const auto& line : lines) {
        bool skip = false;
        for (const auto& d : domains) {
            if (line.find(BLOCK_IP) == 0 && line.find(d) != string::npos) {
                skip = true;
                break;
            }
        }
        if (!skip) newLines.push_back(line);
    }
    if (action == "add") {
        for (const auto& d : domains) {
            newLines.push_back(BLOCK_IP + " " + d);
        }
    }
    writeHosts(newLines);
}

void timerThread(int minutes) {
    this_thread::sleep_for(chrono::minutes(minutes));
    cout << colorize("Time expired. Unblocking...", YELLOW) << endl;
    disableBlocking();
}

void startBlocking(const vector<string>& domains, int minutes) {
    auto root = loadConfig();
    for (const auto& d : domains) {
        root["blocked"].append(d);
    }
    if (minutes > 0) {
        auto now = chrono::system_clock::now();
        time_t tt = chrono::system_clock::to_time_t(now + chrono::minutes(minutes));
        char buf[64];
        strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S", localtime(&tt));
        root["timer_end"] = string(buf);
    }
    root["active"] = true;
    saveConfig(root);
    try {
        updateHosts(domains, "add");
        cout << colorize("Blocked " + to_string(domains.size()) + " domains.", GREEN) << endl;
        if (minutes > 0) {
            cout << colorize("Blocking active for " + to_string(minutes) + " min.", YELLOW) << endl;
            thread(timerThread, minutes).detach();
        }
    } catch (const exception& e) {
        cerr << colorize("Error: " + string(e.what()), RED) << endl;
    }
}

void disableBlocking() {
    auto root = loadConfig();
    vector<string> blocked;
    for (const auto& d : root["blocked"]) blocked.push_back(d.asString());
    if (blocked.empty()) {
        cout << colorize("No active blocking.", YELLOW) << endl;
        return;
    }
    try {
        updateHosts(blocked, "remove");
        root["blocked"] = Json::arrayValue;
        root["active"] = false;
        root["timer_end"] = "";
        saveConfig(root);
        cout << colorize("Blocking disabled.", GREEN) << endl;
    } catch (const exception& e) {
        cerr << colorize("Error: " + string(e.what()), RED) << endl;
    }
}

void showStatus() {
    auto root = loadConfig();
    if (root["active"].asBool() && root.isMember("timer_end") && !root["timer_end"].asString().empty()) {
        string endStr = root["timer_end"].asString();
        tm tm = {};
        stringstream ss(endStr);
        ss >> get_time(&tm, "%Y-%m-%dT%H:%M:%S");
        time_t end = mktime(&tm);
        int remaining = (int)difftime(end, time(nullptr));
        if (remaining > 0) {
            int mins = remaining / 60;
            int secs = remaining % 60;
            cout << colorize("Blocking active. Remaining: " + to_string(mins) + " min " + to_string(secs) + " sec", BLUE) << endl;
        } else {
            cout << colorize("Blocking active but timer expired.", YELLOW) << endl;
        }
    } else if (root["active"].asBool()) {
        cout << colorize("Blocking active (no timer).", BLUE) << endl;
    } else {
        cout << colorize("Blocking inactive.", YELLOW) << endl;
    }
    if (root["blocked"].size() > 0) {
        cout << colorize("Blocked domains (" + to_string(root["blocked"].size()) + "):", GREEN) << endl;
        for (const auto& d : root["blocked"]) {
            cout << "  - " << d.asString() << endl;
        }
    } else {
        cout << colorize("No blocked domains.", YELLOW) << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cout << colorize("Usage: blocker add|remove|list|enable|disable|status|pomodoro|clear [domains...] [options]", YELLOW) << endl;
        return 1;
    }
    string cmd = argv[1];
    vector<string> args;
    for (int i = 2; i < argc; ++i) args.push_back(argv[i]);

    auto root = loadConfig();

    if (cmd == "add") {
        if (args.empty()) {
            cout << colorize("Specify domains to block.", YELLOW) << endl;
            return 1;
        }
        int minutes = 0;
        for (size_t i = 0; i < args.size(); ++i) {
            if (args[i] == "-t" && i+1 < args.size()) {
                minutes = stoi(args[i+1]);
                args.erase(args.begin()+i, args.begin()+i+2);
                break;
            }
        }
        startBlocking(args, minutes);
    }
    else if (cmd == "remove") {
        if (args.empty()) {
            cout << colorize("Specify domains to remove.", YELLOW) << endl;
            return 1;
        }
        vector<string> blocked;
        for (const auto& d : root["blocked"]) blocked.push_back(d.asString());
        for (const auto& d : args) {
            auto it = find(blocked.begin(), blocked.end(), d);
            if (it != blocked.end()) blocked.erase(it);
        }
        root["blocked"] = Json::arrayValue;
        for (const auto& d : blocked) root["blocked"].append(d);
        saveConfig(root);
        try {
            updateHosts(args, "remove");
            cout << colorize("Removed: ", GREEN);
            for (const auto& d : args) cout << d << " ";
            cout << endl;
        } catch (const exception& e) {
            cerr << colorize("Error: " + string(e.what()), RED) << endl;
        }
    }
    else if (cmd == "list") {
        if (root["blocked"].size() == 0) {
            cout << colorize("No blocked domains.", YELLOW) << endl;
        } else {
            cout << colorize("Blocked domains:", GREEN) << endl;
            for (const auto& d : root["blocked"]) {
                cout << "  - " << d.asString() << endl;
            }
        }
    }
    else if (cmd == "enable") {
        if (root["blocked"].size() == 0) {
            cout << colorize("No domains in blocklist. Add them with 'add'.", YELLOW) << endl;
            return 1;
        }
        int minutes = 0;
        for (size_t i = 0; i < args.size(); ++i) {
            if (args[i] == "-t" && i+1 < args.size()) {
                minutes = stoi(args[i+1]);
                break;
            }
        }
        vector<string> blocked;
        for (const auto& d : root["blocked"]) blocked.push_back(d.asString());
        startBlocking(blocked, minutes);
    }
    else if (cmd == "disable") {
        disableBlocking();
    }
    else if (cmd == "status") {
        showStatus();
    }
    else if (cmd == "pomodoro") {
        if (root["blocked"].size() == 0) {
            cout << colorize("No domains in blocklist. Add them with 'add'.", YELLOW) << endl;
            return 1;
        }
        cout << colorize("Starting Pomodoro: 25 minutes work", BLUE) << endl;
        vector<string> blocked;
        for (const auto& d : root["blocked"]) blocked.push_back(d.asString());
        startBlocking(blocked, 25);
        this_thread::sleep_for(chrono::minutes(25));
        cout << colorize("Pomodoro finished! 5 minutes break.", GREEN) << endl;
        disableBlocking();
        this_thread::sleep_for(chrono::minutes(5));
        cout << colorize("Break over. Start next Pomodoro.", BLUE) << endl;
    }
    else if (cmd == "clear") {
        bool force = false;
        for (const auto& arg : args) if (arg == "-f") force = true;
        if (!force) {
            cout << colorize("Clear all blocklist? [y/N] ", YELLOW);
            string ans;
            getline(cin, ans);
            if (ans != "y" && ans != "Y") return 0;
        }
        if (root["blocked"].size() == 0) {
            cout << colorize("Blocklist already empty.", YELLOW) << endl;
            return 0;
        }
        vector<string> blocked;
        for (const auto& d : root["blocked"]) blocked.push_back(d.asString());
        try {
            updateHosts(blocked, "remove");
            root["blocked"] = Json::arrayValue;
            root["active"] = false;
            root["timer_end"] = "";
            saveConfig(root);
            cout << colorize("Blocklist cleared.", GREEN) << endl;
        } catch (const exception& e) {
            cerr << colorize("Error: " + string(e.what()), RED) << endl;
        }
    }
    else {
        cout << colorize("Unknown command: " + cmd, RED) << endl;
    }
    return 0;
}
